testing/test_deploy.py

```diff
@@ -0,0 +1,218 @@
+    # branch at the desired commit.
+    packages = ["kubeflow/core", "kubeflow/tf-serving", "kubeflow/tf-job"]
+
+    for p in packages:
+      util.run(["ks", "pkg", "install", p], cwd=app_dir)
+
+    # Deploy Kubeflow
+    util.run(["ks", "generate", "core", "kubeflow-core", "--name=kubeflow-core",
+              "--namespace=" + namespace.metadata.name], cwd=app_dir)
+
+    util.run(["ks", "apply", "default", "-c", "kubeflow-core",], cwd=app_dir)
+
+    # Verify that the TfJob operator is actually deployed.
+    tf_job_deployment_name = "tf-job-operator"
+    logging.info("Verifying TfJob controller started.")
+    util.wait_for_deployment(api_client, namespace.metadata.name,
+                             tf_job_deployment_name)
+
+    # Verify that JupyterHub is actually deployed.
+    jupyter_name = "tf-hub"
+    logging.info("Verifying TfHub started.")
+    util.wait_for_statefulset(api_client, namespace.metadata.name, jupyter_name)
+
+  main_case = test_util.TestCase()
+  main_case.class_name = "KubeFlow"
+  main_case.name = "deploy-kubeflow"
+  try:
+    test_util.wrap_test(run, main_case)
+  finally:
+    # Delete the namespace
+    logging.info("Deleting namespace %s", namespace_name)
+
+    # We report teardown as a separate test case because this will help
+    # us track down issues with garbage collecting namespaces.
+    teardown = test_util.TestCase(main_case.class_name, "teardown")
+    def run_teardown():
+      core_api = k8s_client.CoreV1Api(api_client)
+      core_api.delete_namespace(namespace_name, {})
+
+    try:
+      test_util.wrap_test(run_teardown, teardown)
+    except Exception as e:  # pylint: disable-msg=broad-except
+      logging.error("There was a problem deleting namespace: %s; %s",
+                    namespace_name, e.message)
+    junit_path = os.path.join(args.test_dir, "junit_kubeflow-deploy.xml")
+    logging.info("Writing test results to %s", junit_path)
+    test_util.create_junit_xml_file([main_case, teardown], junit_path)
+
+def main():  # pylint: disable=too-many-locals
+  logging.getLogger().setLevel(logging.INFO) # pylint: disable=too-many-locals
+  # create the top-level parser
+  parser = argparse.ArgumentParser(
+    description="Test Kubeflow E2E.")
+
+  parser.add_argument(
+    "--test_dir",
+    default="",
+    type=str,
+    help="Directory to use for all the test files. If not set a temporary "
+         "directory is created.")
+
+  parser.add_argument(
+    "--project",
+    default=None,
+    type=str,
+    help="The project to use.")
+
+  parser.add_argument(
+    "--cluster",
+    default=None,
+    type=str,
+    help="The name of the cluster.")
+
+  parser.add_argument(
+    "--zone",
+    default="us-east1-d",
+    type=str,
+    help="The zone for the cluster.")
+
+  parser.add_argument(
+    "--github_token",
+    default=None,
+    type=str,
+    help=("The GitHub API token to use. This is needed since ksonnet uses the "
```

Comment:
Link to https://github.com/ksonnet/ksonnet/blob/master/docs/troubleshooting.md? It was unclear at first why this script needed the github token.

Reply:
Done.